contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,267 @@
+
+        // Send user their requested amount of staked JOE
+        joe.safeTransfer(msg.sender, _amount);
+
+        emit Withdraw(msg.sender, _amount);
+    }
+
+    /// @notice Claim any pending veJOE
+    function claim() external {
+        require(getUserHasNonZeroBalance(msg.sender), "VeJoeStaking: cannot claim any veJOE when no JOE is staked");
+        _claim();
+    }
+
+    /// @notice Get the pending amount of veJOE for a given user
+    /// @param _user The user to lookup
+    /// @return The number of pending veJOE tokens for `_user`
+    function getPendingVeJoe(address _user) public view returns (uint256) {
+        if (!getUserHasNonZeroBalance(_user)) {
+            return 0;
+        }
+
+        UserInfo storage user = userInfos[_user];
+
+        // Calculate amount of pending veJOE based on:
+        // 1. Seconds elapsed since last reward timestamp
+        // 2. Generation rate that the user is receiving
+        // 3. Current amount of user's staked JOE
+        uint256 secondsElapsed = block.timestamp - user.lastRewardTimestamp;
+        if (secondsElapsed == 0) {
+            return 0;
+        }
+
+        // Calculate the generation rate the user should receive (in units of veJOE per sec per JOE).
+        // If the current timestamp is less than or equal to the user's `boostEndTimestamp`,
+        // that means the user is currently receiving boosted benefits so they should receive
+        // `boostedGenerationRate`, otherwise `baseGenerationRate`.
+        uint256 generationRate = block.timestamp <= user.boostEndTimestamp ? boostedGenerationRate : baseGenerationRate;
+
+        uint256 accVeJoePerJoe = secondsElapsed * generationRate;
+
+        uint256 pendingVeJoe = accVeJoePerJoe * user.balance;
+
+        // Get the user's current veJOE balance and maximum veJOE they can hold
+        uint256 userVeJoeBalance = veJoe.balanceOf(_user);
+        uint256 userMaxVeJoeCap = user.balance * maxCap;
+
+        if (userVeJoeBalance < userMaxVeJoeCap) {
+            if (userVeJoeBalance + pendingVeJoe > userMaxVeJoeCap) {
+                return userMaxVeJoeCap - userVeJoeBalance;
+            } else {
+                return pendingVeJoe;
+            }
+        } else {
+            // User already holds maximum amount of veJOE so there is no pending veJOE
+            return 0;
+        }
+    }
+
+    /// @notice Checks to see if a given user currently has staked JOE
+    /// @param _user the user address to check
+    /// @return whether `_user` currently has staked JOE
+    function getUserHasNonZeroBalance(address _user) public view returns (bool) {
+        return userInfos[_user].balance > 0;
+    }
+
+    /// @dev Helper to claim any pending veJOE
+    function _claim() private {
+        uint256 veJoeToClaim = getPendingVeJoe(msg.sender);
+
+        // Update user's last reward timestamp
+        userInfos[msg.sender].lastRewardTimestamp = block.timestamp;
```

Comment:
Should this be updated if the user has 0 vejoe to claim, I think this should be moved inside the `if` statement below.

Reply:
yeah good point